### PR TITLE
Improve how Stream Read and Write Timeouts are changed

### PR DESCRIPTION
**Pull request by nathanwoctopusdeploy**
# Background

Improve the way the read and write stream timeouts are changed for a particular operation.

# Results

This PR does not change any timeouts, just changes the API for how those timeouts are changed.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut/Transport/Protocol/MessageExchangeStream.cs [hunk 1]
```diff
@@ -63,27 +64,29 @@ async Task SendIdentityMessageAsync(string identityLine, CancellationToken cance
...
+                MessageExchangeStreamTimeout.ControlMessageExchangeShortTimeout,
+                async () => await SendControlMessageAsync(Next, cancellationToken));
         }
 
         public async Task SendProceedAsync(CancellationToken cancellationToken)
         {
-            await SendControlMessageAsync(Proceed, cancellationToken);
+            await WithTimeout(
+                MessageExchangeStreamTimeout.ControlMessageExchangeShortTimeout,
+                async () => await SendControlMessageAsync(Proceed, cancellationToken));
```
Comment (sburmanoctopus) [inline source/Halibut/Transport/Protocol/MessageExchangeStream.cs:76]: As we discussed, this is different to what was there before, but we believed that this message does not need the full timeout set, and would make sense to use the lower timeout.

File: source/Halibut/Transport/Protocol/MessageExchangeStreamTimeout.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+using System;
+
+namespace Halibut.Transport.Protocol
+{
+    public enum MessageExchangeStreamTimeout
```
Comment (sburmanoctopus) [inline source/Halibut/Transport/Protocol/MessageExchangeStreamTimeout.cs:5]: 💭 Just a thought, but I wonder if an alternative to this enum is to acknowledge that the send/receive timeouts are a concept, and to make a class instead (which has a send timeout property and a receive timeout property), and replace the 4 `HalibutTimeoutsAndLimits` properties (TcpClientSendTimeout, TcpClientReceiveTimeout, TcpClientHeartbeatSendTimeout, TcpClientHeartbeatReceiveTimeout) with just 2 properties using this new class.

With this remodelling, we would not need this enum, and could simply pass that object into the extension methods below.

Anyway, not vital for this PR, just thought I'd float the idea 😁

**Review by sburmanoctopus — APPROVED** [review]
LGTM, nice one

File: source/Halibut/Transport/Protocol/MessageExchangeStream.cs (hunk 1 above)
Comment (nathanwoctopusdeploy) [reply source/Halibut/Transport/Protocol/MessageExchangeStream.cs:76]: Ahh, good spot. I did not mean for this to make it into this PR so will revert!!

Comment (nathanwoctopusdeploy) [reply source/Halibut/Transport/Protocol/MessageExchangeStream.cs:76]: For a future PR

File: source/Halibut/Transport/Protocol/MessageExchangeStreamTimeout.cs (hunk 2 above)
Comment (nathanwoctopusdeploy) [reply source/Halibut/Transport/Protocol/MessageExchangeStreamTimeout.cs:5]: Yeh that would be a lot cleaner 👍 I'll rework

Comment (nathanwoctopusdeploy) [reply source/Halibut/Transport/Protocol/MessageExchangeStreamTimeout.cs]: I've reworked the change, thoughts @sburmanoctopus

Comment (sburmanoctopus) [reply source/Halibut/Transport/Protocol/MessageExchangeStreamTimeout.cs]: Nice one! LGTM 😁

**Pull request merged**